### PR TITLE
Rework validator exiting, add min balance clause

### PR DESCRIPTION
**Pull request by vbuterin**

### CODE REVIEW

File: specs/beacon-chain.md [hunk 1]
```diff
@@ -560,6 +561,20 @@ def add_validator(validators: List[ValidatorRecord],
         return index
 ```
 
+## Routine for removing a validator
+
+```python
+def exit_validator(index, crystallized_state, penalize, current_slot):
+    v = crystallized_state.validators[index]
```
Comment (djrtwo) [inline specs/beacon-chain.md]: Prefer explicitly named variables in the function defs

 `v` --> `validator`

File: specs/beacon-chain.md [hunk 2]
```diff
@@ -664,18 +679,14 @@ In addition, validators with `status == PENALIZED` lose `B // reward_quotient +
...
+* **[covers logouts]**: If `obj.kind == LOGOUT`, interpret `data[0]` as a validator index as an `int32` and `data[1]` as a signature. If `BLSVerify(pubkey=validators[data[0]].pubkey, msg=hash(LOGOUT_MESSAGE), sig=data[1])`, and `validators[i].status == ACTIVE`, run `exit_validator(data[0], crystallized_state, penalize=False, current_slot=current_slot)`
+* **[covers `NO_DBL_VOTE`, `NO_SURROUND`, `NO_DBL_PROPOSE` slashing conditions]:** If `obj.kind == CASPER_SLASHING`, interpret `data[0]` as a list of concatenated `int32` values where each value represents an index into `validators`, `data[1]` as the data being signed and `data[2]` as an aggregate signature. Interpret `data[3:6]` similarly. Verify that both signatures are valid, that the two signatures are signing distinct data, and that they are either signing the same slot number, or that one surrounds the other (ie. `source1 < source2 < target2 < target1`). Let `indices` be the list of indices in both signatures; verify that its length is at least 1. For each validator index `v` in `indices`, if its `status` does not equal `PENALIZED`, then run `exit_validator(v, crystallized_state, penalize=True, current_slot=current_slot)`
```
Comment (djrtwo) [inline specs/beacon-chain.md]: prefer `current_slot=block.slot`

File: specs/beacon-chain.md [hunk 3]
```diff
@@ -664,18 +679,14 @@ In addition, validators with `status == PENALIZED` lose `B // reward_quotient +
...
 
 * **[covers RANDAO updates]**: If `obj.kind == RANDAO_REVEAL`, interpret `data[0]` as an integer and `data[1]` as a hash32. Set `validators[data[0]].randao_commitment = data[1]`.
 
 #### Finally...
 
+* If the balance of any validator with index `v` is less than `MIN_ONLINE_DEPOSIT_SIZE`, then run `exit_validator(v, crystallized_state, penalize=False, current_slot=current_slot)`
```
Comment (djrtwo) [inline specs/beacon-chain.md]: prefer `current_slot=block.slot`

Comment (djrtwo) [inline specs/beacon-chain.md]: Be explicit about the status required for such a validator (`ACTIVE`)

Something like:
 "For any validator defined by index `v` with `status` equal to `ACTIVE` and `balance` less than `MIN_ONLINE_DEPOSIT_SIZE`, run `exit_validator`..."

**Comment by vbuterin** [comment]
Done the changes.

**Pull request merged**